ArduPlane/soaring.cpp

```diff
@@ -47,175 +47,134 @@ void Plane::update_soaring() {
     }
 
     switch (control_mode->mode_number()) {
-    default:
-        // nothing to do
-        break;
-    case Mode::Number::AUTO:
-    case Mode::Number::FLY_BY_WIRE_B:
-    case Mode::Number::CRUISE:
-        // Test for switch into thermalling mode
-        g2.soaring_controller.update_cruising();
-
-        if (g2.soaring_controller.check_thermal_criteria()) {
-            gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Thermal detected, entering %s", mode_loiter.name());
-            set_mode(mode_loiter, ModeReason::SOARING_THERMAL_DETECTED);
-        }
-        break;
+        default:
+            // nothing to do
+            break;
+        case Mode::Number::AUTO:
+        case Mode::Number::FLY_BY_WIRE_B:
+        case Mode::Number::CRUISE:
+            // Test for switch into thermalling mode
+            g2.soaring_controller.update_cruising();
 
-    case Mode::Number::LOITER: {
-        // Update thermal estimate and check for switch back to AUTO
-        g2.soaring_controller.update_thermalling();  // Update estimate
+            if (g2.soaring_controller.check_thermal_criteria()) {
+                gcs().send_text(MAV_SEVERITY_INFO, "Soaring: Thermal detected, entering %s", mode_loiter.name());
 
-        Vector2f prev_wp, next_wp;
+                set_mode(mode_loiter, ModeReason::SOARING_THERMAL_DETECTED);
+            }
+            break;
 
-        // If previous mode was AUTO and there was a previous NAV command, we can use previous and next wps for drift calculation.
-        if (previous_mode->mode_number() == Mode::Number::AUTO) {
-            AP_Mission::Mission_Command current_nav_cmd = mission.get_current_nav_cmd();
-            AP_Mission::Mission_Command prev_nav_cmd;
+        case Mode::Number::LOITER: {
+            // Update thermal estimate and check for switch back to AUTO
+            g2.soaring_controller.update_thermalling();  // Update estimate
+
+            // Check distance to home.
+            Vector3f position;
+            if (!ahrs.get_relative_position_NED_home(position)) {
+                return;
+            } else if (g2.soaring_controller.max_radius >= 0 &&
+                       powf(position.x,2)+powf(position.y,2) > powf(g2.soaring_controller.max_radius,2) &&
+                    previous_mode->mode_number()!=Mode::Number::AUTO) {
```

Comment:
You don't need to call mode_number() here - you can compare against `&mode_auto` - not that it matters all that much.

Reply:
Good tip, thanks!